streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStore.java

```diff
@@ -0,0 +1,792 @@
+        public byte[] getLatestValue(final Bytes key) {
+            return latestValueStore.get(key);
+        }
+
+        @Override
+        public void putLatestValue(final Bytes key, final byte[] value) {
+            latestValueStore.put(key, value);
+        }
+
+        @Override
+        public void deleteLatestValue(final Bytes key) {
+            latestValueStore.delete(key);
+        }
+
+        @Override
+        public LogicalKeyValueSegment getOrCreateSegmentIfLive(final long segmentId, final ProcessorContext context, final long streamTime) {
+            return segmentStores.getOrCreateSegmentIfLive(segmentId, context, streamTime);
+        }
+
+        @Override
+        public List<LogicalKeyValueSegment> getReverseSegments(final long timestampFrom) {
+            return segmentStores.segments(timestampFrom, Long.MAX_VALUE, false);
+        }
+
+        @Override
+        public long segmentIdForTimestamp(final long timestamp) {
+            return segmentStores.segmentId(timestamp);
+        }
+    }
+
+    private static <T extends VersionedStoreSegment> void doPut(
+        final LatestValueSchema latestValueSchema,
+        final SegmentValueSchema segmentValueSchema,
+        final VersionedStoreClient<T> versionedStoreClient,
+        final Optional<Sensor> expiredRecordSensor,
+        final ProcessorContext context,
+        final long observedStreamTime,
+        final long historyRetention,
+        final Bytes key,
+        final byte[] value,
+        final long timestamp
+    ) {
+        // track the smallest timestamp seen so far that is larger than insertion timestamp.
+        // this timestamp determines, based on all segments searched so far, which segment the
+        // new record should be inserted into. the starting "sentinel timestamp" represents
+        // that the segment should be inserted into the latest value store.
+        long foundTs = SENTINEL_TIMESTAMP;
+
+        // check latest value store
+        PutStatus status = putLatestValueStore(
+            latestValueSchema,
+            segmentValueSchema,
+            versionedStoreClient,
+            context,
+            observedStreamTime,
+            key,
+            value,
+            timestamp,
+            foundTs
+        );
+        if (status.isComplete) {
+            return;
+        } else {
+            foundTs = status.foundTs;
+        }
+
+        // continue search in segments
+        status = putSegments(
+            segmentValueSchema,
+            versionedStoreClient,
+            expiredRecordSensor,
+            context,
+            observedStreamTime,
+            historyRetention,
+            key,
+            value,
+            timestamp,
+            foundTs
+        );
+        if (status.isComplete) {
+            return;
+        } else {
+            foundTs = status.foundTs;
+        }
+
+        // ran out of segments to search. insert into segment as specified by foundTs
+        putFallThrough(
```

Comment:
"fall through" is not very descriptive -- can we find a better name?

I needed to hop between the different method implementation to understand the code. The control flow is not easy (not sure if/how we could improve it right now...)